lib/streamlit/elements/utils.py

```diff
@@ -86,6 +86,40 @@ def check_session_state_rules(
         _shown_default_value_warning = True
 
 
+class CachedWidgetWarning(StreamlitAPIWarning):
+    def __init__(self):
+        super().__init__(
+            """
+Your script uses a widget command in a cached function
+(function decorated with `@st.cache_data` or `@st.cache_resource`).
+This code will only be called when we detect a cache "miss",
+which can lead to unexpected results.
+
+How to fix this:
+* Move all widget commands outside the cached function.
+* Or, if you know what you're doing, use `experimental_allow_widgets=True`
+in the cache decorator to enable widget replay and suppress this warning.
+"""
+        )
+
+
+def check_widget_usage_rules() -> None:
+    """Check if a widget can be correctly used in the current context.
+
+    This currently only checks if a widget is used inside a cached function
+    with `experimental_allow_widgets=False` and shows a warning. But is supposed
+    to be extended to check for other similar cases in the future.
```

Comment:
what other similar cases do you have in mind? if we have such a generic "check" method, it'd be great if you could provide one, two examples of future refactorings that should belong in here. Otherwise I personally would prefer a more concrete method-name

Reply:
Renamed it to `check_cache_replay_rules` with a comment that we should rename this as soon as we can extend it.